### ISSUE TITLE
Start brave in incognito+TOR mode using command line

### ISSUE THREAD

**Issue by edouard-lopez** [brave/brave-browser#2105]
## Description 

**related:** #690,  #1457

Starting Brave in incognito mode via command is possible with the `--incognito` flag. However there is no such flag for the Tor feature.

## Steps to Reproduce

To enable Tor you have to:

   1. Start brave manually or by CLI `brave-browser --incognito` (here in incognito mode) ;
   2. Hit `Alt+Shift+N`  (Linux user) ;
   3. Close previous window.

## Actual result:

Works

## Expected result:

Open directly via a flag, e.g.:

     brave-browser --incognito --tor

## Reproduces how often: 

Every time

## Brave version (brave://version info)

     Version 0.56.12 Chromium: 70.0.3538.77 (Official Build) (64-bit)

**Comment by jackcaron** [comment]
Yes please!!!

Not sure if the right issue also, but when installed on *Ubuntu*, you can lock *Brave* in the launcher. And with a right click, having option for the `incognito` and `incognito+tor` would also be great.

**Comment by jtbrown3** [comment]
Thanks for mentioning this @jackcaron - In Windows 10, this context feature of 'right-clicking' from the TaskBar or Start Menu also exists.  It would be a GREAT if Brave could also add the "**New Incognito+Tor Window**" selection to these special context menus.  This could be an extension of the command-line option being requested here...

**Comment by Brave-Matt** [comment]
Just had a user on Reddit confirm that a way to do this in Windows (tested and works on my end):
https://www.reddit.com/r/brave_browser/comments/a7aflb/how_to_launch_brave_browser_in_incognito_mode_by/
Method:
1. Right-click on Brave Shortcut --> `Properties`
2. In `Target` path, enter `[space]--incognito` after quotations
3. Click `Apply/OK`
4. Brave shortcut launches with Private Window
Note: the `[space]` refers the actual space character - command throws invalid path error without the space between the last `"` and `--incognito`

**Comment by manit-shah** [comment]
**+1** for adding CLI command to open Brave tor window (start Brave --tor perhaps?)
**+1** for adding "New private window with Tor" option to Windows 10 special context menus

**Additional (related) request:**
Allow the user to specify the default new window type (_normal window_, _private window_, or _private window with Tor_). So that a user may choose for Brave to always open to a "private window with Tor". This would be an outstanding addition (which even Chrome does not have).

**Comment by Brave-Matt** [comment]
+1 from reddit: https://www.reddit.com/r/brave_browser/comments/ck9z3v/command_line_option_to_launch_new_private_browser/

**Comment by Brave-Matt** [comment]
+1 https://www.reddit.com/r/brave_browser/comments/csao5r/open_brave_with_tor_by_default_via_shortcut/

**Comment by geradorcpt** [comment]
+1 from reddit:
https://www.reddit.com/r/brave_browser/comments/e6pk3i/windows_run_command_to_open_link_in_private/

**Comment by derycck** [comment]
+1 from reddit:
https://www.reddit.com/r/brave_browser/comments/e6pk3i/windows_run_command_to_open_link_in_private/

**Comment by apenasrr** [comment]
+1 from Reddit:
https://www.reddit.com/r/brave_browser/comments/e6pk3i/windows_run_command_to_open_link_in_private/

**Comment by WhatDothLife** [comment]
Can't wait for this to be implemented.

**Comment by x0rzkov** [comment]
+1 bump

**Comment by RuthlessRuler** [comment]
+1 Bump.

**Comment by cybarspace** [comment]
It's been 2 years already... Please implement it soon. It'd be a killer feature.
Also, please add it to the context menu (.desktop file on Linux) so we get a "New Incognito Window with Tor" option on right-clicking the shortcut.

**Comment by contre95** [comment]
It would be a great feature, yes !

**Comment by kolokd** [comment]
how f*cking long this will take? Do hear me? 
just parse a new cli argument and trigger incognito+tor mode just like incognito.

**Comment by contre95** [comment]
> how f*cking long this will take? Do hear me?
> just parse a new cli argument and trigger incognito+tor mode just like incognito.

Calm down dude !

**Comment by dbjordan1** [comment]
> how f*cking long this will take? Do hear me?
> just parse a new cli argument and trigger incognito+tor mode just like incognito.

Yup, should be pretty simple. Good thing it's open source, we'll keep an eye out for the pull request.

**Comment by woswos** [comment]
If anyone needs this feature for automation (for example, using Brave Incognito+Tor mode in selenium), try using `xdotool windowactivate --sync <Brave's_ window_id>  key alt+shift+n` to send the shortcut to the browser instead. Later, you can use `driver.switch_to.window()` to switch to the Incognito+Tor window and keep running your tests.

If you already thought about using selenium's sendKeys() to open the Incognito+Tor mode, that doesn't work because apparently "keys get sent directly to the render process, bypassing the browser process. So, any keyboard shortcut handlers in the browser process will not be invoked by sendKeys()."

Of course, I would prefer to use `brave-browser --incognito --tor`, but it doesn't exist yet :(

**Comment by Syntist** [comment]
> If anyone needs this feature for automation (for example, using Brave Incognito+Tor mode in selenium), try using `xdotool windowactivate --sync <Brave's_ window_id> key alt+shift+n` to send the shortcut to the browser instead. Later, you can use `driver.switch_to.window()` to switch to the Incognito+Tor window and keep running your tests.
> 
> If you already thought about using selenium's sendKeys() to open the Incognito+Tor mode, that doesn't work because apparently "keys get sent directly to the render process, bypassing the browser process. So, any keyboard shortcut handlers in the browser process will not be invoked by sendKeys()."
> 
> Of course, I would prefer to use `brave-browser --incognito --tor`, but it doesn't exist yet :(

I can't seem to get focus on new window, i execute xdotool through subprocess!

**Comment by russeg** [comment]
Since it's been 2 years, I'm assuming this is a very very hard problem. Is there an unsurmountable technical reason why It has not been implemented? It seems it's not as simple as parsing `--tor` or whatever and launching a new browser window, like some people suggested? Am I wrong about this?

**Comment by dbjordan1** [comment]
I believe this is the hardest problem the team has ever encountered.
</troll>

Doug Jordan

**Comment by 141** [comment]
It seems impossible for them as Tor is not part of the browser itself, but a built-in extension. If they can merge Tor in, then it would be simple.

**Comment by xmha97** [comment]
You can use
```
brave onion
```

**Comment by vikaspatelp83** [comment]
Create a shortcut with the target: 
**"C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe" onion**
it will fire brave tor.

**Comment by austinwk** [comment]
> You can use:
> 
> ```
> brave onion
> ```

Thanks for the tip! For some users, like myself, it may be `brave-browser onion`. I'm on Kubuntu and used the official apt install instructions (not snap).

**Comment by jtbrown3** [comment]
> Create a shortcut with the target:
> `"C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe" onion`
> it will fire brave tor.

This works partially, but on my system, Brave also tries to load a new "onion/" page in a new tab and fails with _"This site can't be reached"_ error.

Is there a more specific syntax that can be used so the failed new tab doesn't open?

**Comment by darkwiz666** [comment]
> > Create a shortcut with the target:
> > `"C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe" onion`
> > it will fire brave tor.
> 
> @vikaspatelp83, This works partially, but on my system, Brave also tries to load a new "onion/" page in a new tab and fails with _"This site can't be reached"_ error.
> 
> Is there a more specific syntax that can be used so the failed new tab doesn't open?

I'm bumping this because of all the crap that's been suggested this is the one thing that is nearly working on Windows. adding "--" does nothing but open a normal window (and a single dash does nothing, obviously)

**Comment by anonymous-ne0** [comment]
As of now, I just use:
```
Set WshShell = WScript.CreateObject("WScript.Shell")
WshShell.AppActivate "brave"
WshShell.SendKeys "+%{n}"
```
as a .vbs file, which is executed by the batch file (after the brave window is opened) 😐

**Comment by anonymous-ne0** [comment]
> As of now, I just use:
> 
> ```
> Set WshShell = WScript.CreateObject("WScript.Shell")
> WshShell.AppActivate "brave"
> WshShell.SendKeys "+%{n}"
> ```
> 
> as a .vbs file, which is executed by the batch file (after the brave window is opened) 😐

The batch file is: 
```
start brave --incognito
timeout 2
wscript brave_tor_keystroke.vbs
```
The timeout of 2s is for the window to load (might be more/less for you)

**Comment by anonymous-ne0** [comment]
The batch file is:
```
start brave --incognito
timeout 2
wscript [path/filename of vbs file].vbs
```
The timeout of 2s is for the window to load (might be more/less for you)

**Comment by anonymous-ne0** [comment]
> As of now, I just use:
> 
> ```
> Set WshShell = WScript.CreateObject("WScript.Shell")
> WshShell.AppActivate "brave"
> WshShell.SendKeys "+%{n}"
> ```
> 
> as a .vbs file, which is executed by the batch file (after the brave window is opened) 😐

The batch file is:
```
start brave --incognito
timeout 2
wscript [path/filename of vbs file].vbs
```
The timeout of 2s is for the window to load (might be more/less for you)

**Comment by darkwiz666** [comment]
> > > Create a shortcut with the target:
> > > `"C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe" onion`
> > > it will fire brave tor.
> > 
> > 
> > @vikaspatelp83, This works partially, but on my system, Brave also tries to load a new "onion/" page in a new tab and fails with _"This site can't be reached"_ error.
> > Is there a more specific syntax that can be used so the failed new tab doesn't open?
> 
> After getting the brave widow, just open a new tab and search anything this new window is opened in tor browser mode.

**That's not what is being asked:** we're trying to do this WITHOUT the GUI. External programs can't invoke the Tor-mode browser (properly), even though there is a flag to do so.

**Comment by vikaspatelp83** [comment]
> > > > Create a shortcut with the target:
> > > > `"C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe" onion`
> > > > it will fire brave tor.
> > > 
> > > 
> > > @vikaspatelp83, This works partially, but on my system, Brave also tries to load a new "onion/" page in a new tab and fails with _"This site can't be reached"_ error.
> > > Is there a more specific syntax that can be used so the failed new tab doesn't open?
> > 
> > 
> > After getting the brave widow, just open a new tab and search anything this new window is opened in tor browser mode.
> 
> **That's not what is being asked:** we're trying to do this WITHOUT the GUI. External programs can't invoke the Tor-mode browser (properly), even though there is a flag to do so.

Yes that's right.

**Comment by DavidFromNL** [comment]
> > You can use:
> > ```
> > brave onion
> > ```
> 
> Thanks for the tip! For some users, like myself, it may be `brave-browser onion`. I'm on Kubuntu and used the official apt install instructions (not snap).

Running "brave-browser onion" on my Brave 1.19.88 / Fedora 33 / x86_64 system gets my a hanging browser, followed by a segmentation fault every time. Also, opening https://3g2upl4pq6kufc4m.onion/ (Duckduckgo) from the command line does the same, no matter if brave was already running or not. Opening Duck by typing said address in an already-running brave browser is no problem at all.

**Comment by Fablo020** [comment]
I got the /Onion can't be reached error too.
After enabling "Automatically redirect .onion sites" in settings, it is working now. But a native solution like -onion start parameter would be way better IMO. Thank for updating Tor anyways finally I can skip the captchas and block elements using uBlock.
Just FYI I used: "brave.exe onion" - I tested it with any Site like "brave.exe 3g2upl4pq6kufc4m.onion" It opens 2 tabs in Tor one "onion/" and another with the Link, in this case: https://3g2upl4pq6kufc4m.onion/.

**Comment by Brave-Matt** [comment]
+1 Community:
https://community.brave.com/t/setting-private-or-private-with-tor-as-default-tab-types/235252/5
@rebron this issue has been around for some time -- is there any way we can get someone to take a stab at this or is it blocked by something?

**Comment by M09NST3R** [comment]
Related question, is there some flag (brave://flags) that forces Brave to always use Tor mode?

**Comment by jdnixx** [comment]
> After enabling "Automatically redirect .onion sites" in settings, it is working now. But a native solution like -onion start parameter would be way better IMO.
Hey thanks, this is a good workaround. Look forward to a real CLI option though glad to see this just recently got put on the radar

**Comment by bridiver** [comment]
I think we need some more detail here for `Actual result`. What should happen if there is a previously saved profile that would normally be on startup? I think this should be a separate profile or maybe even a separate user data dir, but we need to outline these scenarios for QA and browser tests.

**Comment by bridiver** [comment]
it's possible that `--incognito` already handles ^^, but we still need to explain what the expected behavior is

**Comment by NullHypothesis** [comment]
This feature was implemented in https://github.com/brave/brave-core/pull/9533. It makes it possible to start Brave in Tor mode by running:
```
brave-browser --tor
```
Note that if the `--tor` switch is used together with `--silent-launch`, Tor won't be launched.

**Comment by rebron** [comment]
Closing. Fixed per comment above.

**Comment by xmha97** [comment]
> brave-browser --tor

This command doesn't work on Windows.
![image](https://user-images.githubusercontent.com/23485114/128094026-991d1f9e-9134-4e59-a1e4-39669bbf1550.png)
Please help me or reopen this issue.

**Comment by stephendonner** [comment]
Verified `PASSED` using the testplan from https://github.com/brave/brave-core/pull/9533 with 


Brave | 1.29.48 Chromium: 92.0.4515.131 (Official Build) nightly (x86_64)
-- | --
Revision | 6b8d6c56ce21e38a72f7c4becb5abc1fa5134f29-refs/branch-heads/4515@{#1933}
OS | macOS Version 11.5.1 (Build 20G80)

## Steps:

<details>
<summary>Test 1</summary>

1. Open a website in browser window 1 (`abc.com`) and another website in browser window 2 (`bbc.com`)
2. Quit Brave and launch it with `--tor` command line flag
3. Open a normal window by pressing Ctrl + n or by clicking on Menu -> New window
4. Windows from step 1 should be restored

Confirmed `abc.com` and `bbc.com` reopened

<img width="910" alt="Screen Shot 2021-08-03 at 3 52 53 PM" src="https://user-images.githubusercontent.com/387249/128096953-0840ee9c-bfa6-410e-96ce-673e47892b5c.png">

</details>

<details>

<summary>Test 2</summary>

1. Open a website in browser window 1 (`abc.com`) and another website in browser window 2 (`bbc.com`)
2. Quit Brave and launch it with `--tor` command line flag
3. Quit Brave
4. Launch Brave without `--tor`
5. Windows from step 1 should be restored

Confirmed `abc.com` and `bbc.com` reopened

<img width="911" alt="Screen Shot 2021-08-03 at 3 50 28 PM" src="https://user-images.githubusercontent.com/387249/128097004-916bf6a7-227e-4e02-923e-94a6adc70f23.png">

</details>

<details>
<summary>Test 3</summary>

1. Open a website in browser window 1 (`abc.com`) and another website in browser window 2 (`bcc.com`)
2. Crash browser using `chrome://inducebrowsercrashforrealz/`
3. Launch Brave with `--tor`
4. Open a normal window by pressing Ctrl + n or by clicking on Menu -> New window
5. Windows from step 1 should be restored

Confirmed `abc.com` and `bbc.com` reopened
example | example
---------|---------
<img width="1312" alt="Screen Shot 2021-08-03 at 3 43 45 PM" src="https://user-images.githubusercontent.com/387249/128097060-5197bdf0-c512-4f0a-8750-5a98e0d99fc1.png"> | <img width="910" alt="Screen Shot 2021-08-03 at 3 52 53 PM" src="https://user-images.githubusercontent.com/387249/128097095-2ff520f7-1d3f-429a-84b2-80131077a6cb.png">

</details>

<details>

<summary>Test 4</summary>

1. Open a website in browser window 1 (`abc.com`) and another website in browser window 2 (`bbc.com`)
2. Crash browser using `chrome://inducebrowsercrashforrealz/`
3. Launch Brave with `--tor`
4. Quit Brave
5. Launch Brave without `--tor`
6. Windows from step 1 should be restored

Confirmed `abc.com` and `bbc.com` reopened

<img width="908" alt="Screen Shot 2021-08-03 at 3 54 15 PM" src="https://user-images.githubusercontent.com/387249/128097132-7f136bc5-da11-477c-8702-b93d309c4db9.png">

</details>

<details>
<summary>Test 5</summary>


1. Launch Brave with `--tor`
2. Visit `https://check.torproject.org/api/ip`
3. Make sure that the page contains the string `"IsTor":true`.

Confirmed I saw `{"IsTor":true,"IP":"185.220.102.249"}`

<img width="1139" alt="Screen Shot 2021-08-03 at 3 45 52 PM" src="https://user-images.githubusercontent.com/387249/128096900-28e636a1-9c37-4f4d-8c5c-4a96113ba4ef.png">
</details>

**Comment by stephendonner** [comment]
> > brave-browser --tor
> 
> This command doesn't work on the stable version.
> ![image](https://user-images.githubusercontent.com/23485114/128094026-991d1f9e-9134-4e59-a1e4-39669bbf1550.png)
> Please help me or reopen this issue.

hi @xmha97 - this does work in `1.29.48` and newer `nightly` builds, and will be in the August 31st release (https://github.com/brave/brave-browser/wiki/Brave-Release-Schedule), if not sooner (if we take the change and uplift/back-port to `1.28.x` branch).

**Comment by gnfpt** [comment]
Brave doesn't start in tor mode when app mode is also used.

ie:
  `brave --tor --app="https://whatismyipaddress.com/"`

~/gnf.pt

**Comment by MatiasMontanna** [comment]
+1. I support this. This is already available to use as command-line argument? --tor. Doesn't work for me.
